### ISSUE TITLE
Problems with empty strings

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap#2695]
The following removes spurious spaces from a string:

![image](https://user-images.githubusercontent.com/2138696/93898434-c93c2200-fd25-11ea-83da-000fba894f47.png)

And yet this doesn't

![image](https://user-images.githubusercontent.com/2138696/93898318-aa3d9000-fd25-11ea-8f38-2f5d35c18e39.png)

Despite them being equal

![image](https://user-images.githubusercontent.com/2138696/93898543-e7a21d80-fd25-11ea-9573-cc23b10062c9.png)

and even identical

![image](https://user-images.githubusercontent.com/2138696/93898809-3c459880-fd26-11ea-947b-7900be32a76f.png)

What is going on? And how should one do this?

**Comment by jmoenig** [comment]
I always test for length:
![filter_out_spaces](https://user-images.githubusercontent.com/2596285/93970443-691d9e00-fd6e-11ea-8290-381de9b66763.png)

**Comment by ToonTalk** [comment]
That is nicer. But join() is acting strange:

![image](https://user-images.githubusercontent.com/2138696/94029124-10421a00-fdef-11ea-8933-6cad99bfa536.png)
![image](https://user-images.githubusercontent.com/2138696/94029189-20f29000-fdef-11ea-8bbd-62d11c4dff1e.png)

**Comment by DarDoro** [comment]
![for topgamerpro16 script pic (1)](https://user-images.githubusercontent.com/16109071/94034252-166dd180-fdc2-11ea-980b-93b880d9da87.png)
![for topgamerpro16 script pic](https://user-images.githubusercontent.com/16109071/94034274-1b328580-fdc2-11ea-9657-f79e6048446a.png)

![for topgamerpro16 script pic (2)](https://user-images.githubusercontent.com/16109071/94034348-330a0980-fdc2-11ea-92cc-abc773d50452.png)

Yes "join()" seems greedy so its not good candidate for null anymore ;(
![for topgamerpro16 script pic (3)](https://user-images.githubusercontent.com/16109071/94035897-fa6b2f80-fdc3-11ea-9a70-582baaac276d.png)

**Comment by brianharvey** [comment]
Join is weird because it makes a text form of things that aren't text, e.g., lists and procedures.  I'm guessing what you're reporting is part of the same weirdness.

**Comment by DarDoro** [comment]
![untitled script pic (12)](https://user-images.githubusercontent.com/16109071/94057057-627b3f00-fddf-11ea-9852-a6b9f25e66ca.png)

It seems that a variadic parameter without an empty slot is considered unbound rather than empty. It works this way even in custom blocks so it's probably an intended but not well documented feature.

**Comment by jmoenig** [comment]
nah, JS functions are always called with the current process as very last argument, so you can write your own interpolated blocks.

**Comment by DarDoro** [comment]
It was an unfortunate choice to use JavaScript function as an example.
Most important was that this code
![untitled script pic (13)](https://user-images.githubusercontent.com/16109071/94066661-10411a80-fded-11ea-8213-466d146dfa27.png)
acts as
![untitled script pic (14)](https://user-images.githubusercontent.com/16109071/94066674-17682880-fded-11ea-92ca-e92c016bdbdb.png)

**Comment by ToonTalk** [comment]
But this means that you can't substitute equals (or even identicals) in a script without completely changing its behaviour. See the first example in this issue.

**Comment by jmoenig** [comment]
Naw, it only means that Snap! doesn't have the notion of `nil`

**Comment by DarDoro** [comment]
> ![for topgamerpro16 script pic (1)](https://user-images.githubusercontent.com/16109071/94034252-166dd180-fdc2-11ea-980b-93b880d9da87.png)
instead of join<>

**Comment by ToonTalk** [comment]
I'd like to see Brian translate join() (of no arguments) to Scheme such that it can be used in `map`.

Why should

![image](https://user-images.githubusercontent.com/2138696/94131681-ce69b000-fe90-11ea-8994-41dbdb5c322a.png)

behave the same as

![image](https://user-images.githubusercontent.com/2138696/94131608-babe4980-fe90-11ea-8d28-947f71dd1c08.png)

?

**Comment by ToonTalk** [comment]
And note that these behave differently as I would expect:

![image](https://user-images.githubusercontent.com/2138696/94132196-85fec200-fe91-11ea-8a54-b025b4795e6a.png)

**Comment by ToonTalk** [comment]
> It was an unfortunate choice to use JavaScript function as an example.
> Most important was that this code
> ![untitled script pic (13)](https://user-images.githubusercontent.com/16109071/94066661-10411a80-fded-11ea-8213-466d146dfa27.png)
> acts as
> ![untitled script pic (14)](https://user-images.githubusercontent.com/16109071/94066674-17682880-fded-11ea-92ca-e92c016bdbdb.png)

@DarDoro - you are correct about `join` and` append` but not `list`.

Can someone explain this behavior? 

Jens's "Naw, it only means that Snap! doesn't have the notion of nil" does explain it.
